### PR TITLE
Add SyncObjective method to Wallet

### PR DESCRIPTION
**Pull request by lalexgap**
# Description
Add a new SyncObjective method to the wallet API.

This PR introduces a new method on the wallet `syncObjectives`. It accepts a list of `objectiveId`s and returns a message that can be sent to other counterparties. 

The syncObjectives method will also call `syncChannel` for any relevant channels so the message payload will contain all relevant channels for the objective.

The main use case this is designed around is for resending objectives if a message gets dropped.

## How Has This Been Tested? 

Two new test files have been added to test the functionality. One is a unit test, while the other is more of an integration style test with two wallets.

It would be nice to test this is out on an e2e style test however I think the amount of effort to do that is outside the scope of this PR.

### Code quality
- [x] I have written clear commit messages
- [x] I have performed a self-review of my own code
- [x] This change does not have an unduly wide scope
- [x] I have separated logic changes from refactor changes (formatting, renames, etc.)
- [x] I have commented my code wherever necessary (can be 0)
- [x] I have added tests that prove my fix is effective or that my feature works, if necessary
### Project management
- [x] I have applied the [appropriate labels](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#b304e56734a74dfbb341b8b4b27b1c0c)
- [x] I have linked to all relevant issues (can be 0)
- [x] I have added all dependent tickets (can be 0)
- [x] I have assigned myself to this PR
- [x] I have chosen the appropriate pipeline on zenhub

### CODE REVIEW

File: packages/server-wallet/src/wallet/wallet-response.ts [hunk 1]
```diff
@@ -86,7 +86,9 @@ export class WalletResponse {
     myIndex: number,
     participants: Participant[]
   ): void {
-    this.createdObjectives.push(objective);
+    // TODO: The objective model currently always sets the participants to an empty array when creating the DBObjective
```
Comment (lalexgap) [inline packages/server-wallet/src/wallet/wallet-response.ts:89]: Currently `toObjective` always sets the `participants` to an [empty array](https://github.com/statechannels/statechannels/blob/517ed467a5dbc542a89aee837669817a7ccafd3d/packages/server-wallet/src/models/objective.ts#L204).

**Comment by andrewgordstewart** [comment]
> This PR introduces a new method on the wallet syncObjectives. It accepts a list of objectiveIds and returns a message that can be sent to other counterparties.
> 
> The main use case this is designed around is for resending objectives if a message gets dropped.

I think the scope of this function should be larger. If my channel manager (CM) expects an objective to make progress, and I call `syncObjective`, I should expect it to make progress.

I think the change here is a good start, pending a review.

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  database: 'TEST_B',
+});
+
+let a: Wallet;
+let b: Wallet;
+
+beforeAll(async () => {
+  await Promise.all([DBAdmin.dropDatabase(aWalletConfig), DBAdmin.dropDatabase(bWalletConfig)]);
+  await DBAdmin.createDatabase(aWalletConfig);
+  await DBAdmin.createDatabase(bWalletConfig);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:28]: Is this copied from some other file? If so, seems like the wrong pattern to _create & drop_ some databases within multiple files.

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+afterAll(async () => {
+  await Promise.all([a.destroy(), b.destroy()]);
+  await Promise.all([DBAdmin.dropDatabase(aWalletConfig), DBAdmin.dropDatabase(bWalletConfig)]);
+});
+
+test('Objectives can be synced if a message is lost', async () => {
+  const createChannelParams: CreateChannelParams = createChannelArgs();
+
+  // We mimic not receiving a message containing objectives
+  const messageToLose = await a.createChannel(createChannelParams);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:48]: [required] What's missing in this test suite is what happens when the message here is not lost, but `syncObjectives` is called.

It's possible that this is already resolved by some other test case. Is there a test that checks what happens when an objective is pushed multiple times?

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+});
+
+async function getObjective(knex: Knex, objectiveId: string): Promise<DBObjective | undefined> {
+  const model = await ObjectiveModel.query(knex).findById(objectiveId);
+
+  if (model) {
+    return model.toObjective();
+  } else {
+    return undefined;
+  }
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:78]: I think this works?
```suggestion
  return model?.toObjective();
```

File: packages/server-wallet/src/wallet/store.ts [hunk 5]
```diff
@@ -280,6 +280,14 @@ export class Store {
     return {states: vars.map(ss => _.merge(ss, channelConstants)), channelState};
   }
 
+  async getObjectivesByIds(objectiveIds: string[]): Promise<DBObjective[]> {
+    return ObjectiveModel.forIds(objectiveIds, this.knex);
+  }
+
+  async getChannelsByIds(channelIds: string[]): Promise<ChannelState[]> {
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/store.ts:287]: This function is added in this PR, but it seems unrelated to the rest of the PR, and it's not used elsewhere. Is that intentional?

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      .withGraphFetched('funding');
+    const objective = await ObjectiveModel.insert(openChannelObjective(), testKnex);
+
+    const syncResult = await wallet.syncObjectives([objective.objectiveId]);
+    // TODO: Currently the objective model differs slightly from what we see on the message
+    const massagedObjective = {
+      ...objective,
+      participants: [aliceP(), bob()],
+    };
+    expect(syncResult.newObjectives).toEqual([massagedObjective]);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:52]: Why should the objective be a `newObjective` after calling `syncObjective`? It's not a new objective, right?

File: packages/server-wallet/src/wallet/wallet-response.ts (hunk 1 above)
Comment (andrewgordstewart) [reply packages/server-wallet/src/wallet/wallet-response.ts:89]: I think we need some consult to @geoknee here. I do not know the reason why `participants` is an empty array. What is the effect of adding `participants` to it?

File: packages/server-wallet/src/wallet/wallet.ts [hunk 7]
```diff
@@ -287,6 +287,38 @@ export class SingleThreadedWallet
...
+        {objectiveIds: missingObjectives},
+        'The following objectives cannot be found'
+      );
+      throw new Error('Could not find all objectives');
+    }
+
+    for (const o of objectives) {
+      const channel = await this.store.getChannelState(o.data.targetChannelId);
+      // This will make sure any relevant channel information is synced
+      await this._syncChannel(channel.channelId, response);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/wallet.ts:313]: 👍

File: packages/server-wallet/src/wallet/wallet.ts [hunk 8]
```diff
@@ -287,6 +287,38 @@ export class SingleThreadedWallet
...
+
+      const {participants} = channel;
+      response.queueCreatedObjective(o, channel.myIndex, participants);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/wallet.ts:316]: I find this confusing, since the objective wasn't just created. It might even be _incorrect_, since an app might read `newObjectives`, which is read from `response.createdObjectives` and assume that they are new objectives.

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    expect(syncResult.channelResults).toEqual([targetChannel.channelResult]);
+
+    expect((syncResult.outbox[0].params.data as any).objectives).toEqual([
+      expect.objectContaining(_.pick(objective, 'data')),
+    ]);
+  });
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:58]: There's no expectation in this test on the effect of `syncChannel`. I think there should be one, right?

**Review by andrewgordstewart — CHANGES_REQUESTED** [review]
I'm not ready to approve this yet, because
1. I'm not yet sure what the effect of calling `syncObjective` is, when the message is not dropped
2. Marking objectives as `createdObjectives` after a `syncObjective` call seems incorrect.
3. The test should also check that channels get synced, rigth?

**Comment by andrewgordstewart** [comment]
> > This PR introduces a new method on the wallet syncObjectives. It accepts a list of objectiveIds and returns a message that can be sent to other counterparties.
> > The main use case this is designed around is for resending objectives if a message gets dropped.
> 
> I think the scope of this function should be larger. If my channel manager (CM) expects an objective to make progress, and I call `syncObjective`, I should expect it to make progress.
> 
> I think the change here is a good start, pending a review.

I see now that in the implementation, `syncChannel` is also triggered. This could be listed in the description, as I believe your intent is to `squash & merge`, and that is probably important information to be in the commit message.

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts (hunk 9 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:58]: I'm not sure I understand? The only responsibility of `syncChannel` is to generate a message that can be sent to the other participants. Which is why the assertions are on the `syncResult`

Comment (andrewgordstewart) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:58]: But the assertions do not check for the messages that are generated by `syncChannel`. Or, am I misreading the tests?

File: packages/server-wallet/src/wallet/wallet.ts (hunk 8 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/wallet.ts:316]: I think this is due to my misunderstanding of the `newObjectives`.  I thought that this was communicated to other counterparties instead of the `app`. Oops!

File: packages/server-wallet/src/wallet/wallet-response.ts (hunk 1 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/wallet-response.ts:89]: While I think we should figure that out, this change is no longer needed for this PR and will be removed.

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 2 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:28]: Yup, this was copied from [here](https://github.com/statechannels/statechannels/blob/master/packages/server-wallet/src/__test__/create-and-close-channel/direct-funding.test.ts)

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts (hunk 9 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:58]: It does by checking the `(syncResult.outbox[0].params.data as any).objectives`. However, this could be more clear/thorough so I will clean this up!

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 2 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:28]: This probably should be a category of tests with its own `jest.config` that setups and tears down these databases.

File: packages/server-wallet/src/wallet/store.ts (hunk 5 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/store.ts]: Fixed in 7b17570

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 4 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts]: Resolved in 360b5df

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts (hunk 9 above)
Comment (andrewgordstewart) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:58]: `syncChannels` generates other messages that are not in the `objectives` field. There are no assertions specifying that these messages are generated.

Comment (lalexgap) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:52]: Ah sorry I misread the initial method in your comment 🤦

File: packages/server-wallet/src/wallet/wallet-response.ts [hunk 10]
```diff
@@ -110,6 +99,21 @@ export class WalletResponse {
...
+   * - notifying the app
+   */
+  queueCreatedObjective(
+    objective: DBObjective,
+    myIndex: number,
+    participants: Participant[]
+  ): void {
+    this.createdObjectives.push(objective);
+
+    this.queueSendObjective(objective, myIndex, participants);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/wallet-response.ts:114]: Clever!

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 11]
```diff
@@ -68,6 +68,82 @@ test('Objectives can be synced if a message is lost', async () => {
...
+      recipient: 'bob',
+      sender: 'alice',
+      data: {
+        signedStates: [expect.objectContaining({turnNum: 0})],
+        requests: [],
+      },
+    },
+  });
+  expect(newObjectives).toHaveLength(0);
+  expect(channelResults).toHaveLength(0);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:101]: These expectations should have the same behaviour no matter if you `b.pushMessage` or not, right?. The interesting thing to test is _what happens when `b` receives the objective again_, right?

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 12]
```diff
@@ -68,6 +68,82 @@ test('Objectives can be synced if a message is lost', async () => {
...
+  // The only result will be a message containing the latest states
+  expect(outbox).toHaveLength(1);
+  expect(outbox[0]).toMatchObject({
+    method: 'MessageQueued',
+    params: {
+      recipient: 'bob',
+      sender: 'alice',
+      data: {
+        signedStates: [expect.objectContaining({turnNum: 0})],
+        requests: [],
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:96]: ## Real comment
Thinking through this, why should `b.pushMessage` change the behaviour of `a.syncObjective` in this case? Did somehow `a` learn that `b` has the objective in this test case?

## Meta comment
It's a bit sad that to understand the point of this test, I need to understand that
1. `requests: []` is the important thing to focus on
2. `expect(['foo']).toMatchObject([]);` fails

File: packages/server-wallet/src/wallet/store.ts [hunk 13]
```diff
@@ -284,10 +284,6 @@ export class Store {
     return ObjectiveModel.forIds(objectiveIds, this.knex);
   }
 
-  async getChannelsByIds(channelIds: string[]): Promise<ChannelState[]> {
-    return (await Channel.forIds(channelIds, this.knex)).map(c => c.protocolState);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/store.ts]: `Channel.forIds` was also introduced in this PR, and is now no longer used.

File: packages/server-wallet/src/wallet/wallet.ts (hunk 8 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/wallet.ts]: Fixed in d39fc70

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts (hunk 6 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts]: Fixed in d39fc70

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 11 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:101]: Oops, you're correct. Will correct.

File: packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts (hunk 9 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/__test__/sync-objectives.test.ts:75]: This has been updated in 103e7e0. We now more thoroughly check the `outbox`

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 12 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts]: Resolved in e6a1605

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 11 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts]: Resolved in e6a1605

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 3 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:48]: Updated in 44611f7

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+  expect(getChannelResultFor(channelId, resultB0.channelResults)).toMatchObject({
+    status: 'proposed',
+    turnNum: 0,
+  });
+});
+
+test('handles the objective being synced even if no message is lost', async () => {
+  const createChannelParams: CreateChannelParams = createChannelArgs();
+
+  // We mimic not receiving a message containing objectives
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:74]: This comment seems to be incorrect in this context.

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+      data: {
+        signedStates: [expect.objectContaining({turnNum: 0})],
+      },
+    },
+  });
+  expect(newObjectives).toHaveLength(0);
+  expect(channelResults).toHaveLength(1);
+});
+
+test('Can successfully push the sync objective message multiple times', async () => {
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:109]: This test describes pushing the sync objective message multiple times, but it seems like it was only pushed once. Am I missing something?

File: packages/server-wallet/src/models/channel.ts [hunk 16]
```diff
@@ -152,7 +152,14 @@ export class Channel extends Model implements ChannelColumns {
   };
 
   static jsonAttributes = ['vars', 'participants', 'initialSupport'];
-
+  static async forIds(channelIds: Bytes32[], txOrKnex: TransactionOrKnex): Promise<Channel[]> {
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/models/channel.ts:155]: This method is added in this PR, but it is not used elsewhere. Is there a reason it's included here?

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 15 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:109]: We push on line 129 and line 131 but I could add comments making that clearer.

Comment (andrewgordstewart) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:109]: yes, my apologies, I missed that.

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+  // We would then call sync after some time of waiting and not making progress
+  const syncResult = await a.syncObjectives([objectiveId]);
+
+  // We should not see the objective yet
+  expect(await getObjective(b.knex, objectiveId)).toBeUndefined();
+
+  // We should be able to push
+  await b.pushMessage(getPayloadFor(bob().participantId, syncResult.outbox));
+
+  const {outbox, newObjectives, channelResults} = await b.pushMessage(
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:131]:
```suggestion
  // We assert on the resulting values if we push a 2nd time.
  const {outbox, newObjectives, channelResults} = await b.pushMessage(
```

File: packages/server-wallet/src/models/channel.ts (hunk 16 above)
Comment (lalexgap) [reply packages/server-wallet/src/models/channel.ts]: resolved in efa6f61

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 17 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:131]: I've added a similar comment in 5f20aa2

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 14 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts]: resolved in 5f20aa2

File: packages/server-wallet/src/__test__/sync-objectives.test.ts [hunk 18]
```diff
@@ -0,0 +1,152 @@
+import {CreateChannelParams} from '@statechannels/client-api-schema';
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/__test__/sync-objectives.test.ts:1]: nit: this file ought to be called something a bit different -- otherwise, it's not obvious why it's separate from the test file included below.

I think there's a set of tests which play two wallets against each other. Perhaps they should be grouped under `__test__/with-peers/...` or something like that.

File: packages/server-wallet/src/wallet/wallet.ts [hunk 19]
```diff
@@ -288,6 +288,38 @@ export class SingleThreadedWallet
...
+    if (missingObjectives.length > 0) {
+      this.logger.error(
+        {objectiveIds: missingObjectives},
+        'The following objectives cannot be found'
+      );
+      throw new Error('Could not find all objectives');
+    }
+
+    for (const o of objectives) {
+      const channel = await this.store.getChannelState(o.data.targetChannelId);
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/wallet.ts:312]: Do we really need to fetch the `channel` twice? (`_syncChannel` also fetches the channel.)

When I look at the implementation of `_syncChannel`, it appears difficult to pull this off.

**Review by andrewgordstewart — APPROVED** [review]
I'm glad this was simpler than it was thought to be!

I worry it's going to get more complex once wallets communicate whether they've accepted or rejected the objectives, but that's ok for now.

File: packages/server-wallet/src/__test__/sync-objectives.test.ts (hunk 18 above)
Comment (lalexgap) [reply packages/server-wallet/src/__test__/sync-objectives.test.ts:1]: I agree! I think that could be done under this issue: https://github.com/statechannels/statechannels/issues/3279

File: packages/server-wallet/src/wallet/wallet.ts (hunk 19 above)
Comment (lalexgap) [reply packages/server-wallet/src/wallet/wallet.ts:312]: I think we could do this with some refactoring but I'm not sure it's worth the effort right now?

Comment (andrewgordstewart) [reply packages/server-wallet/src/wallet/wallet.ts:312]: Agreed.

I think it's worth marking this, either with a comment or an issue. I'll leave that up to you.

Comment (lalexgap) [reply packages/server-wallet/src/wallet/wallet.ts:315]: Added a comment in https://github.com/statechannels/statechannels/pull/3275/commits/8158ea764c92be74378850d52591b4d28fb627df
